Sources/AsyncAlgorithms/zip/Zip2StateMachine.swift

```diff
@@ -0,0 +1,519 @@
+//
+//  Zip2StateMachine.swift
+//  
+//
+//  Created by Thibault Wittemberg on 13/09/2022.
+//
+
+final class Zip2Runtime<Base1: AsyncSequence, Base2: AsyncSequence>
+where Base1: Sendable, Base1.Element: Sendable, Base2: Sendable, Base2.Element: Sendable {
+  private let stateMachine = Zip2StateMachine<Base1.Element, Base2.Element>()
+
+  init(_ base1: Base1, _ base2: Base2) {
+    let task = Task { [stateMachine] in
```

Comment:
it might make things a bit more correct if the task is started off in the first call to next instead of an initializer. Generally speaking we should probably consider the behavior when tasks to spawn iteration are created - I think ideally it should always be on-demand; else there may be some edge case work that is done out side of expectations down stream.

Reply:
Yeah we should decide on that, I have been going with creating the Tasks in the inits of the Iterators. I believe that is slightly better since it kicks off the Task setup a bit earlier instead of making the first call to next more expensive. Arguably the iterator might never be used but I think it you do that then you don’t mind the Tasks being setup

Reply:
If we postulate that next is a form of demand then no work should be done until the demand is applied. Therefore we should avoid task creation until it is requested; it can looked at as; if you call makeAsyncIterator do we expect computation to be done? I think the answer should be no.

Reply:
Thanks a lot @phausler, @FranzBusch for this first round of review. I will address everything related to value type semantic and side effect free state machine. As for the start of the task, I leave that decision to you of course. Is there a comparison point that we could find with how Combine handles things between what is done when subscribing and what is done when requesting the next values ?

Reply:
Hi @FranzBusch, @phausler I've addressed most of the comments. We still have to take position on "when" do we start the upstream tasks

Reply:
@phausler @FranzBusch The task now begin at the first `next` call (same as Debounce). Before doing the zip3 version I want to make sure I got this right.

Thanks a lot.

Reply:
Yea the rule comes to this; all appreciable work for an AsyncSequence should happen in the next of the iterator. This is because that next signifies demand to do things. If someone calls makeAsyncIterator that should do as little as possible as per performance impact.

Reply:
Ok so you think it is ready for a zip3 implementation

Reply:
Sure, and to be honest you can feel free to do that as it's own PR and we can merge this.

Reply:
@phausler @FranzBusch I've pushed the zip3 version in the same PR.
Perhaps @FranzBusch would have some improvements to suggest performance wise.
Thanks.

Reply:
Just left a larger review here. The key point for me is that we can definitely implement both `zip2` and `zip3` with a single state machine and storage. I did the same thing for `merge`. I think this is important to do since we want to be able to support a variadic generics implementation at some point and the current approach doesn't really work with that in mind.